### PR TITLE
Remove EntityFramework dependency from HealthChecks.UI.Core

### PR DESCRIPTION
**Pull request by Freakazoid182**
**What this PR does / why we need it**:

Removes the `Microsoft.EntityFrameworkCore` dependency from `HealthChecks.UI.Core`

Introducing a new `HealthChecks.UI.Data` library to share code for libraries that require the EF models.

This allows for applications to use the `HealthChecks.UI.Client` library without pulling in an EF dependency. There's no reason for applications using the `HealthChecks.UI.Client` library forcibly update their EF version.

**Which issue(s) this PR fixes**:

Relates to #747

**Special notes for your reviewer**:

I'm not 100% about the versioning of the libraries. Technically this a breaking change. Is there anything I did wrong on versioning? Or is this updated through the release cycle?

**Does this PR introduce a user-facing change?**:

No

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [X] Code compiles correctly
- [X] Created/updated tests (none new introduced)
- [X] Unit tests passing
- [X] End-to-end tests passing
- [X] Extended the documentation (no specific documentation on this topic available AFIAK)
- [X] Provided sample for the feature (does not apply)

### CODE REVIEW

**Comment by sungam3r** [comment]
@Freakazoid182 Please rebase onto `master` if you are goint to continue work on this PR.

**Comment by Freakazoid182** [comment]
> @Freakazoid182 Please rebase onto `master` if you are going to continue work on this PR.

This has been done now

**Comment by sungam3r** [comment]
I see many conflicted files. Perhaps side-effect from #1175 .

**Comment by Freakazoid182** [comment]
> I see many conflicted files. Perhaps side-effect from #1175 .

Resolved

File: src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(NetFrameworkVersion)</TargetFramework>
+    <PackageTags>$(PackageTags);Data</PackageTags>
+    <Description>HealthChecks.UI.Data package containing builder and model definitions</Description>
```
Comment (sungam3r) [inline src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj:6]: Need to change description in HealthChecks.UI.Core.csproj

File: src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    <Version>$(HealthCheckUIData)</Version>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.5" />
```
Comment (sungam3r) [inline src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj:11]:
```suggestion
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.10" />
```

File: src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    <TargetFramework>$(NetFrameworkVersion)</TargetFramework>
+    <PackageTags>$(PackageTags);Data</PackageTags>
+    <Description>HealthChecks.UI.Data package containing builder and model definitions</Description>
+    <Version>$(HealthCheckUIData)</Version>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.10" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="6.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks" Version="6.0.5" />
```
Comment (sungam3r) [inline src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj:13]:
```suggestion
    <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks" Version="6.0.10" />
```

File: src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    <PackageTags>$(PackageTags);Data</PackageTags>
+    <Description>HealthChecks.UI.Data package containing builder and model definitions</Description>
+    <Version>$(HealthCheckUIData)</Version>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.10" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="6.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks" Version="6.0.10" />
+    <PackageReference Include="System.Text.Json" Version="6.0.4" />
```
Comment (sungam3r) [inline src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj:14]:
```suggestion
    <PackageReference Include="System.Text.Json" Version="6.0.6" />
```

File: src/HealthChecks.UI/package-lock.json [hunk 5]
```diff
@@ -2707,8 +2707,7 @@
       "version": "3.0.0",
       "resolved": "https://registry.npmjs.org/normalize-path/-/normalize-path-3.0.0.tgz",
       "integrity": "sha512-6eZs5Ls3WtCisHWp9S2GUy8dqkpGi4BVSz3GaqiE6ezub0512ESztXUwUB6C6IKbQkY2Pnb/mD4WYojCRwcwLA==",
-      "dev": true,
```
Comment (sungam3r) [inline src/HealthChecks.UI/package-lock.json:2710]: Did you intentionally change that file?

File: src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(NetFrameworkVersion)</TargetFramework>
+    <PackageTags>$(PackageTags);Data</PackageTags>
+    <Description>HealthChecks.UI.Data package containing builder and model definitions</Description>
+    <Version>$(HealthCheckUIData)</Version>
```
Comment (sungam3r) [inline src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj:7]: Forgot to add this property into `dependencies.props`.

Comment (sungam3r) [reply src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj:7]: Also you need to append new package in `healthchecks_ui_*` 3 yaml fiels.

**Comment by sungam3r** [comment]
@Freakazoid182 I reviewed this one. It was really easy indeed, I'm sorry for very long delay. I have few comments.

> move UIHealthReport from UI.Core to UI.Client

What about this note from #747 ?

> I'm not 100% about the versioning of the libraries. Technically this a breaking change. Is there anything I did wrong on versioning? Or is this updated through the release cycle?

Yes, technically this a breaking change. Each package from this repo **can** be published with independent version but in practice versions differ only by minor/patch. See `dependencies.props` file. Maybe we can just bump each version  to 7.0 in the light of upcoming in November NET7 release. I'm fine either way, I reviewed your changes and they look good to me so I would like to  merge after fixings issues. Also I should note that I don't use any ~.UI package from this repo. ping @carlosrecuero @unaizorrilla @CarlosLanderas for comments.

**Comment by Freakazoid182** [comment]
@sungam3r Thank you for reviewing and making some fixes.

I've made updates following your remarks.

I've looked at the [comment](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/issues/747)
>  move UIHealthReport from UI.Core to UI.Client

I think this is possible, although the `UIHealthStatus` enum is currently defined in `UIHealthReport.cs`. So that will need to be extracted out into a separate file. Please let me know what you prefer, I think keeping `UIHealthReport` in `UI.Core` is also fine.

https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/dd9140718c9d16b7d04bf7253488bae837012a9e/src/HealthChecks.UI.Core/UIHealthReport.cs#L74

I've set the version for the `HealthChecks.UI.Data` project to `6.0.5` in line with the other UI libraries. I'm sure the version will be updated before it's released?

File: .github/workflows/healthchecks_ui_cd_preview.yml [hunk 7]
```diff
@@ -39,6 +41,8 @@ jobs:
       run: dotnet build --no-restore ./src/HealthChecks.UI/HealthChecks.UI.csproj -c $BUILD_CONFIG
     - name: Build UI.Client
       run: dotnet build --no-restore ./src/HealthChecks.UI.Client/HealthChecks.UI.Client.csproj -c $BUILD_CONFIG
+    - name: Build UI.Data
```
Comment (sungam3r) [inline .github/workflows/healthchecks_ui_cd_preview.yml:44]: + pack & publish

File: src/HealthChecks.UI/package-lock.json [hunk 8]
```diff
@@ -2932,8 +2930,7 @@
       "version": "2.2.2",
       "resolved": "https://registry.npmjs.org/picomatch/-/picomatch-2.2.2.tgz",
       "integrity": "sha512-q0M/9eZHzmr0AulXyPwNfZjtwZ/RBZlbN3K3CErVrk50T2ASYI7Bye0EvekFY3IP1Nt2DHu0re+V2ZHIpMkuWg==",
-      "dev": true,
-      "optional": true
+      "dev": true
```
Comment (sungam3r) [inline src/HealthChecks.UI/package-lock.json:2933]:
```suggestion
      "dev": true,
      "optional": true
```

File: src/HealthChecks.UI/package-lock.json [hunk 9]
```diff
@@ -2707,8 +2707,6 @@
       "version": "3.0.0",
       "resolved": "https://registry.npmjs.org/normalize-path/-/normalize-path-3.0.0.tgz",
       "integrity": "sha512-6eZs5Ls3WtCisHWp9S2GUy8dqkpGi4BVSz3GaqiE6ezub0512ESztXUwUB6C6IKbQkY2Pnb/mD4WYojCRwcwLA==",
```
Comment (sungam3r) [inline src/HealthChecks.UI/package-lock.json:2709]:
```suggestion
      "integrity": "sha512-6eZs5Ls3WtCisHWp9S2GUy8dqkpGi4BVSz3GaqiE6ezub0512ESztXUwUB6C6IKbQkY2Pnb/mD4WYojCRwcwLA==",
      "dev": true,
      "optional": true
```

**Comment by sungam3r** [comment]
> I think keeping UIHealthReport in UI.Core is also fine.

OK. Not in the scope of this PR.

> I've set the version for the HealthChecks.UI.Data project to 6.0.5 in line with the other UI libraries. I'm sure the version will be updated before it's released?

Use version that you would like to see it published.

File: build/dependencies.props [hunk 10]
```diff
@@ -55,13 +55,14 @@
...
     <HealthChecksUIPostgreSQLStorage>6.0.4</HealthChecksUIPostgreSQLStorage>
     <HealthCheckSystem>6.0.5</HealthCheckSystem>
-    <HealthCheckUI>6.0.5</HealthCheckUI>
-    <HealthCheckUIClient>6.0.5</HealthCheckUIClient>
-    <HealthCheckUICore>6.0.5</HealthCheckUICore>
-    <HealthCheckUIInMemoryStorage>6.0.5</HealthCheckUIInMemoryStorage>
-    <HealthCheckUIMySqlStorage>6.0.5</HealthCheckUIMySqlStorage>
-    <HealthCheckUISQLiteStorage>6.0.5</HealthCheckUISQLiteStorage>
-    <HealthCheckUISqlServerStorage>6.0.5</HealthCheckUISqlServerStorage>
+    <HealthCheckUI>7.0.0</HealthCheckUI>
```
Comment (sungam3r) [inline build/dependencies.props:58]: I would not bump to 7 in light of upcoming migration of all packages to NET7.

**Comment by sungam3r** [comment]
@Freakazoid182 #1574 introduces a lot of changes preparing repo to v7 packages including EF core version bump from 6 to 7.

**Comment by sungam3r** [comment]
@Freakazoid182 I'm fine to merge after pulling changes from `master`.

File: src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <PackageTags>$(PackageTags);Data</PackageTags>
+    <Description>HealthChecks.UI.Data package containing data models and database context definition</Description>
+    <VersionPrefix>$(HealthCheckUIData)</VersionPrefix>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="7.0.0" />
```
Comment (sungam3r) [inline src/HealthChecks.UI.Data/HealthChecks.UI.Data.csproj:11]:
```suggestion
```

**Comment by sungam3r** [comment]
@Freakazoid182 I reviewed this one again and I'm ready to merge after fixing sln conflict.

**Comment by sungam3r** [comment]
Thanks.

**Comment by Freakazoid182** [comment]
Thanks for reviewing and merging!

**Pull request merged**